substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+		/// The number of tickets allowed to be submitted in one call is equal to the epoch length.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit_tickets(tickets.len() as u32))]
+		pub fn submit_tickets(
+			origin: OriginFor<T>,
+			tickets: BoundedVec<TicketEnvelope, EpochLengthFor<T>>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+
+			debug!(target: LOG_TARGET, "Received {} tickets", tickets.len());
+
+			let epoch_length = T::EpochLength::get();
+			let current_slot_idx = Self::current_slot_index();
+			if current_slot_idx > epoch_length / 2 {
+				warn!(target: LOG_TARGET, "Tickets shall be submitted in the first epoch half",);
+				return Err("Tickets shall be submitted in the first epoch half".into())
+			}
+
+			let Some(verifier) = RingVerifierData::<T>::get().map(|v| v.into()) else {
+				warn!(target: LOG_TARGET, "Ring verifier key not initialized");
+				return Err("Ring verifier key not initialized".into())
+			};
+
+			let next_authorities = Self::next_authorities();
+
+			// Compute tickets threshold
+			let next_config = Self::next_config().unwrap_or_else(|| Self::config());
+			let ticket_threshold = sp_consensus_sassafras::ticket_id_threshold(
+				next_config.redundancy_factor,
+				epoch_length as u32,
+				next_config.attempts_number,
+				next_authorities.len() as u32,
+			);
+
+			// Get next epoch params
+			let randomness = NextRandomness::<T>::get();
+			let epoch_idx = EpochIndex::<T>::get() + 1;
+
+			let mut valid_tickets = BoundedVec::with_bounded_capacity(tickets.len());
+
+			for ticket in tickets {
+				debug!(target: LOG_TARGET, "Checking ring proof");
+
+				let Some(ticket_id_output) = ticket.signature.outputs.get(0) else {
+					debug!(target: LOG_TARGET, "Missing ticket vrf output from ring signature");
+					continue
+				};
+				let ticket_id_input =
+					vrf::ticket_id_input(&randomness, ticket.body.attempt_idx, epoch_idx);
+
+				// Check threshold constraint
+				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_output);
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Ignoring ticket over threshold ({:032x} >= {:032x})", ticket_id, ticket_threshold);
+					continue
+				}
+
+				// Check for duplicates
+				if TicketsData::<T>::contains_key(ticket_id) {
+					debug!(target: LOG_TARGET, "Ignoring duplicate ticket ({:032x})", ticket_id);
+					continue
+				}
+
+				// Check ring signature
+				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
+				if ticket.signature.ring_vrf_verify(&sign_data, &verifier) {
+					TicketsData::<T>::set(ticket_id, Some(ticket.body));
+					valid_tickets
+						.try_push(ticket_id)
+						.expect("Input segment has same length as bounded destination vector; qed");
+				} else {
+					debug!(target: LOG_TARGET, "Proof verification failure for ticket ({:032x})", ticket_id);
+				}
+			}
+
+			if !valid_tickets.is_empty() {
+				Self::append_tickets(valid_tickets);
+			}
```

Comment:
We should not accept any tx that is having invalid tickets. Or is there any reason?